test/expr/unary/keypath/keypath.swift

```diff
@@ -1006,9 +1006,10 @@ func testMemberAccessOnOptionalKeyPathComponent() {
   // expected-error@-1 {{value of optional type '(Int, Int)?' must be unwrapped to refer to member '0' of wrapped base type '(Int, Int)'}}
   // expected-note@-2 {{use unwrapped type '(Int, Int)' as key path root}}{{4-15=(Int, Int)}}
   
-  // TODO(diagnostics) Improve diagnostics refering to key path root not able to be infered as an optional type.
-  SR5688_KP(\.count)
-  // expected-error@-1 {{value of optional type 'String?' must be unwrapped to refer to member 'count' of wrapped base type 'String'}}
+  SR5688_KP(\.count) // expected-error {{key path root inferred as optional type 'String?' must be chained to refer to member 'count' of unwrapped type 'String'}}
+  // expected-note@-1 {{chain the optional using '?.' to access unwrapped type member 'count'}} {{15-15=?.}}
+  let _ : KeyPath<String?, Int> = \.count // expected-error {{key path root inferred as optional type 'String?' must be chained to refer to member 'count' of unwrapped type 'String'}}
+  // expected-note@-1 {{chain the optional using '?.' to access unwrapped type member 'count'}} {{37-37=?.}}
```

Comment:
In this case, the fix-it should probably be `!.` because the result type is not `Optional`. But the current compiler has some known bugs in handing `.!.` in keypath expressions.

Reply:
Reported! 
https://bugs.swift.org/browse/SR-13442

Reply:
I don't think we should suggest using 'unsafelyUnwrapped'. From [docs](https://developer.apple.com/documentation/swift/optional/1641793-unsafelyunwrapped):

> The unsafelyUnwrapped property provides the same value as the forced unwrap operator (postfix !). However, in optimized builds (-O), no check is performed to ensure that the current instance actually has a value. Accessing this property in the case of a nil value is a serious programming error and could lead to undefined behavior or a runtime error.

Reply:
It's the same for `!` I think... both will throw a runtime error on a key path application like
```swift
func f<T>(_ x: KeyPath<String?, T>) -> T {
  let s: String? = nil
  return s[keyPath: x] // Thread 1: Fatal error: Unexpectedly found nil while unwrapping an Optional value
}

_ = f(\.!.count) // same for f(\.unsafelyUnwrapped.count)
```

Reply:
In debug mode, sure. But in release mode (`-O`), it can lead to undefined behavior since no check will be performed before accessing the value. Using postfix `!` does a check even in `-O`.

The docs also mention (in red background!):

> This property trades safety for performance. Use unsafelyUnwrapped only when you are confident that this instance will never be equal to nil and only after you’ve tried using the postfix ! operator.

Reply:
Ah ok, but at the end, I think the safest suggestion is `?.`, as we can provide both(as other cases) fix-its `?.` and `!.` since the later right now crashes the compiler, we could let only the optional chain? cc @rintaro

Reply:
Just added both `?.` and `!.` fix-its